hub-cloud-upload/README.md

```diff
@@ -0,0 +1,30 @@
+# hub-cloud-upload
+
+
+This hubverse action uploads hub data to the cloud. Currently, the workflow has a single job, `upload`,
+that pushes data to an AWS S3 bucket.
+
+The `upload` job perform the following steps:
+
+1. Inspect the hub's admin config (`admin.json`) for a `cloud` group.
+2. If `cloud.enabled` is set to `true`:
+    - authenticate to AWS
+    - use `cloud.host.storage` to determine the name of the hub's S3 bucket
+    - sync the hub's `hub-config`, `model-metadata`, and `model-output` directories to the S3 bucket
+
+**Note**: This action is safe to use with non cloud-enabled hubs. 
+If the hub's `admin.config` does not contain a `cloud` group or has `cloud.enabled` set to anything other than `true`,
+the action will skip AWS-related steps.
+
+
+## AWS setup
+
+Before using this action, a member of the hubverse development team will need to "onboard" the hub to AWS. Onboarding is
+a one-time process that creates:
+
+- An AWS S3 bucket for the hub
+- A set of AWS permissions that allow the repo's GitHub workflows to write to the bucket
+
+**Important**: The repo's write permissions are limited to the `main` branch. Running this action on another branch
+or on a fork will fail.
+
```

Comment:
Excellently-timed question--I've been working on some info for next week's dev meeting to talk through setup for cloud-enabled hubs (if you're on Confluence, you can see the [WIP here](https://reichlab.atlassian.net/wiki/spaces/RLD/pages/9207810/2024-03-05+Dev+Meeting)).

In short, each hub gets an AWS bucket _and_ an AWS IAM role that allows writing to the bucket. The role's trust policy states that it can only be used (_assumed_ in AWS parlance) by a request originating from GitHub. That same policy further clarifies that the GitHub request must be associated with a specific repo and branch.

So at a high level:

1. The GitHub action tells AWS that it wants to "assume" a role. You can see that request in [this line of `hubverse-aws-upload.yaml`](https://github.com/Infectious-Disease-Modeling-Hubs/hubverse-actions/blob/bsweger/add-cloud-sync-action/hubverse-aws-upload/hubverse-aws-upload.yaml#L47).
2. AWS receives the request and check's the role's trust policy.
3. If everything checks out, AWS issues a temporary token to GitHub, which allows GitHub to use the role (this is why we don't need to store any AWS creds as GitHub secrets)

For example, this is the trust policy used by the test `hubverse-infrastructure` repo.

```
1.   {
2.      "Version": "2012-10-17",
3.      "Statement": [
4.          {
5.              "Effect": "Allow",
6.             "Principal": {
7.                  "Federated": "arn:aws:iam::767397675902:oidc-provider/token.actions.githubusercontent.com"
8.              },
9.              "Action": "sts:AssumeRoleWithWebIdentity",
10.             "Condition": {
11.                 "StringEquals": {
12.                     "token.actions.githubusercontent.com:aud": "sts.amazonaws.com",
13.                     "token.actions.githubusercontent.com:sub": "repo:Infectious-Disease-Modeling-Hubs/hubverse-infrastructure:ref:refs/heads/main"
14.                 }
15.             }
16.         }
17.     ]
18. }
```

* line 7 says that the GitHub trusted identity provider in our AWS account (which I set up a few weeks ago) is eligible to request the use of this trust policy
* line 9 lists the single action that is permitted under this trust policy: [assuming the role via a web identity](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRoleWithWebIdentity.html) (in this case, the web identity is GitHub)
* line 12 states that the AWS resource requested by GitHub must be AWS's STS (secure token service). The AWS `aws-actions/configure-aws-credentials` action sets this by default, so you won't see that coded explicitly in `hubvserse-aws-upload.yaml`)
* line 13 says that the request from GitHub must originate from the main branch of `Infectious-Disease-Modeling-Hubs/hubverse-infrastructure`